core/module/module.go

```diff
@@ -1,57 +1,63 @@
 
 import (
 	"context"
 	"encoding/json"
+	"strings"
 	"time"
 
-	"github.com/odpf/entropy/core/resource"
 	"github.com/odpf/entropy/pkg/errors"
 )
 
-// Module is responsible for achieving desired external system states based
-// on a resource in Entropy.
-type Module interface {
-	// Plan SHOULD validate the action on the current version of the resource,
-	// return the resource with config/status/state changes (if any) applied.
-	// Plan SHOULD NOT have side effects on anything other than the resource.
-	Plan(ctx context.Context, spec Spec, act ActionRequest) (*Plan, error)
-
-	// Sync is called repeatedly by Entropy core until the returned state is
-	// a terminal status. Module implementation is free to execute an action
-	// in a single Sync() call or split into steps for better feedback to the
-	// end-user about the progress.
-	// Sync can return state in resource.StatusDeleted to indicate resource
-	// should be removed from the Entropy storage.
-	Sync(ctx context.Context, spec Spec) (*resource.State, error)
+type Store interface {
+	GetModule(ctx context.Context, urn string) (*Module, error)
+	ListModules(ctx context.Context, project string) ([]Module, error)
+	CreateModule(ctx context.Context, m Module) error
+	UpdateModule(ctx context.Context, m Module) error
+	DeleteModule(ctx context.Context, urn string) error
 }
 
-// Plan represents the changes to be staged and later synced by module.
-type Plan struct {
-	Resource      resource.Resource
-	ScheduleRunAt time.Time
+// Module represents all the data needed to initialize a particular module.
+type Module struct {
+	URN       string    `json:"urn"`
+	Name      string    `json:"name"`
+	Project   string    `json:"project"`
+	Spec      Spec      `json:"spec"`
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
 }
 
-// Spec represents the context for Plan() or Sync() invocations.
 type Spec struct {
-	Resource     resource.Resource             `json:"resource"`
-	Dependencies map[string]ResolvedDependency `json:"dependencies"`
-}
-
-type ResolvedDependency struct {
-	Kind   string          `json:"kind"`
-	Output json.RawMessage `json:"output"`
+	Configs json.RawMessage `json:"configs"`
 }
 
 // Descriptor is a module descriptor that represents supported actions, resource-kind
 // the module can operate on, etc.
 type Descriptor struct {
-	Kind         string            `json:"kind"`
-	Actions      []ActionDesc      `json:"actions"`
-	Dependencies map[string]string `json:"dependencies"`
-	Module       Module            `json:"-"`
+	Kind          string                                     `json:"kind"`
+	Actions       []ActionDesc                               `json:"actions"`
+	Dependencies  map[string]string                          `json:"dependencies"`
+	DriverFactory func(conf json.RawMessage) (Driver, error) `json:"-"`
+}
+
+func (mod *Module) Validate(isCreate bool) error {
```

Comment:
Is there any specific reason we are mixing the logic of validation and initialization here? It looks like there are 2 behaviours in validate. Wouldn't it better to separate them?

Perhaps `Validate` should doing any mutation and just validating the struct and we could have another func called `Init` or `Build` that mutates the struct itself?

Reply:
mostly because they are never used separately and also, the sanitise part should ideally be called before validate in almost all cases. the responsibility here is mostly `make it valid`.. we could call it `sanitiseAndValidate()` or just `sanitise()` as well. The error is coming because if certains fields have no sane-defaults and are missing, there is no way to sanitise them without explicit value from user.

Reply:
I see, With `Validate` name,  I am under the impression that it only does validation, so having a more explicit name maybe would be more helpful.

Reply:
yep, i agree. ideally Validate() should not be mutating anything and should just be checking the validity.. i will change the naming.